src/main/java/org/marketdesignresearch/mechlib/outcomerules/OutcomeRuleScaler.java

```diff
@@ -0,0 +1,60 @@
+package org.marketdesignresearch.mechlib.outcomerules;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.marketdesignresearch.mechlib.core.Allocation;
+import org.marketdesignresearch.mechlib.core.BidderAllocation;
+import org.marketdesignresearch.mechlib.core.BidderPayment;
+import org.marketdesignresearch.mechlib.core.Outcome;
+import org.marketdesignresearch.mechlib.core.Payment;
+import org.marketdesignresearch.mechlib.core.bid.bundle.BundleValueBids;
+import org.marketdesignresearch.mechlib.core.bidder.Bidder;
+import org.marketdesignresearch.mechlib.instrumentation.MipInstrumentation;
+import org.marketdesignresearch.mechlib.outcomerules.ccg.constraintgeneration.PotentialCoalition;
+
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+public class OutcomeRuleScaler implements OutcomeRule{
+	private final BigDecimal scale;
+	private final BundleValueBids<?> originalBids;
+	private final OutcomeRule base;
+	
+	@Override
+	public void setMipInstrumentation(MipInstrumentation mipInstrumentation) {
+		this.base.setMipInstrumentation(mipInstrumentation);	
+	}
+	@Override
+	public Outcome getOutcome() {
+		Outcome originalOutome = this.base.getOutcome();
+		// create Payment with original values
+		Payment newPayment = new Payment(originalOutome.getPayment().getPaymentMap().entrySet().stream()
+					.collect(Collectors.toMap(Map.Entry::getKey, 
+							e->new BidderPayment(
+									e.getValue().getAmount().divide(this.scale,RoundingMode.HALF_UP))))
+					,originalOutome.getPayment().getMetaInfo());
+		
+		// create Allocation with original values
+		Map<Bidder,BidderAllocation> tradesMap = new LinkedHashMap<>();
+		for(Map.Entry<Bidder,BidderAllocation> e :originalOutome.getAllocation().getTradesMap().entrySet()) {
+			BigDecimal value = e.getValue().getAcceptedBids().stream()
+					.map(b -> originalBids.getBid(e.getKey()).getBidForBundle(b.getBundle()).getAmount())
+					.reduce(BigDecimal.ZERO,BigDecimal::add);
+			tradesMap.put(e.getKey(), new BidderAllocation(value, e.getValue().getBundle(), e.getValue().getAcceptedBids()));
+		}
+		
+		Set<PotentialCoalition> potentialCoalitions = null;
+		if(originalOutome.getAllocation().getPotentialCoalitions() != null) {
+			potentialCoalitions = originalOutome.getAllocation().getPotentialCoalitions().stream().map(p -> new PotentialCoalition(p.getBundle(), p.getBidder(), p.getValue().divide(this.scale,RoundingMode.HALF_UP))).collect(Collectors.toSet());
+		}
+		
+		Allocation newAllocation = new Allocation(tradesMap, originalBids, originalOutome.getMetaInfo(), potentialCoalitions);
+		return new Outcome(newPayment, newAllocation);
+	}
+}

```

Comment:
Can you please deeply review the upscaling of the CCG outcome? I think this is the part where a mistake is most likely.

Reply:
It looks good to me. Also, I wrote a test with large numbers that verify that the scaling works appropriately. I formatted it and added a little catch for dividing BigDecimal values - we need to explicitly set a scale for the result, otherwise if the number is too small, it rounds to zero, leading to wrong results. I actually thought throwing an error in this case is appropriate, because if you need to scale further than 1e-10 down, your result will not be usable anyway I think since CPLEX does some rounding that may render the upscaled results invalid. Feel free to handle this differently, though.

Reply:
Thats a good point. I added the same logic to the BidBasedWinnerDetermination as well.